scarlet/model.py

```diff
@@ -0,0 +1,118 @@
+from abc import ABC, abstractmethod
+
+from .parameter import Parameter
+
+
+class Model(ABC):
+    """Model base class.
+
+    This class stores and provides access to parameters and sub-ordinate models.
+
+    Parameters
+    ----------
+    parameters: list of `~scarlet.Parameter`
+    children: list of `~scarlet.Model`
+        Subordinate models.
+    """
+
+    def __init__(self, *parameters, children=None):
+
+        if len(parameters) == 0:
+            self._parameters = ()
+        elif isinstance(parameters, Parameter):
+            self._parameters = (parameters,)
+        elif isinstance(parameters, (list, tuple)):
+            for p in parameters:
+                assert isinstance(p, Parameter)
+            self._parameters = parameters
+        else:
+            raise TypeError(
+                "parameter must be None, a Parameter, or a list of Parameters"
+            )
+
+        if children is None:
+            children = ()
+        if hasattr(children, "__iter__"):
+            for c in children:
+                assert isinstance(c, Model)
+            self._children = children
+        else:
+            assert isinstance(children, Model)
+            self._children = tuple(children)
+
+        self.check_parameters()
+
+    @property
+    def parameters(self):
+        """List of parameters, including from the children.
+        """
+        return self._parameters + tuple(p for c in self.children for p in c.parameters)
```

Comment:
I don't think that's necessary any more. The list of parameters is always complete (incl. from the children), so there's no ambiguity.

Reply:
I didn't mean in terms of ambiguity, I meant because the list of parameters should rarely change, so it would be faster to cache it on init as opposed to adding tuples together every time `parameters` is called.

Reply:
I don't like caching the result of an operation that takes very little time

Reply:
True, but we call `self.parameters` enough that I feel like it might add up. The only time I don't like caching things like this is when we expect something to change regularly, since we would spend more time caching than we would retrieving it. In this case we are unlikely to ever modify the pointer to `_parameters` for a model or its children, so a simple change doesn't cost us anything.

Reply:
I timed the call in the quickstart notebook with 9 sources. It takes 60 µs, comparable to a single application of monotonicity. I think we are OK.

Reply:
Fair enough, I withdraw my objection.